hal/spi_api.h

```diff
@@ -19,220 +18,331 @@
- */
-int spi_master_block_write(spi_t *obj, const char *tx_buffer, int tx_length, char *rx_buffer, int rx_length, char write_fill);
-
-/** Check if a value is available to read
+ * @param[in,out] obj   This.
+ * @param[in,out] args  A pointer to a spi_transfer_args_t object.
  *
- * @param[in] obj The SPI peripheral to check
- * @return non-zero if a value is available
+ * @return True on success.
  */
-int  spi_slave_receive(spi_t *obj);
+bool spi_transfer(spi_t *obj, spi_transfer_args_t *args);
 
-/** Get a received value out of the SPI receive buffer in slave mode
+/**
+ * Schedules a transfer using the given parameters.
+ * @warning All parameters and the embedded references must stay "alive" until completion of the operation.
  *
- * Blocks until a value is available
- * @param[in] obj The SPI peripheral to read
- * @return The value received
- */
-int  spi_slave_read(spi_t *obj);
-
-/** Write a value to the SPI peripheral in slave mode
+ * @param[in,out] obj       This.
+ * @param[in,out] args      A pointer to a spi_transfer_args_t object.
+ * @param[in]     context   A context to be used by the callback.
+ * @param[in]     cb        A callback invoked upon completion of this transaction.
  *
- * Blocks until the SPI peripheral can be written to
- * @param[in] obj   The SPI peripheral to write
- * @param[in] value The value to write
+ * @return A reference counted handle to the transaction.
  */
-void spi_slave_write(spi_t *obj, int value);
+spi_async_handle_t* spi_async_transfer(
+    spi_t *obj,
+    spi_transfer_args_t *args,
+    void *context,
+    spi_event_f cb);
 
-/** Checks if the specified SPI peripheral is in use
+/**
+ * Aborts the transaction referenced by the spi_async_handle_t.
+ *
+ * This function notifies the lowlevel implementation that the given transaction must be cancelled
+ * (if not already started) or aborted (if currently running) and returns.
+ *
+ * The callback associated with this spi_async_handle_t will be invoked with SPI_EVENT_TYPE_ON_ABORT
+ * as its evtype argument.
  *
- * @param[in] obj The SPI peripheral to check
- * @return non-zero if the peripheral is currently transmitting
+ * This does nothing when used on an already cancelled/aborted transaction.
+ *
+ * @param[in] handle This.
+ *
+ * The handle is consumed in this operation and should no longer be used.
  */
-int  spi_busy(spi_t *obj);
+void spi_async_abort(spi_async_handle_t* handle);
 
-/** Get the module number
+/**
+ * Tells the low-level driver that the upper layer is no longer keeping this handle.
+ *
+ * This does nothing if passed a NULL pointer.
  *
- * @param[in] obj The SPI peripheral to check
- * @return The module number
+ * @param[in] handle This.
  */
-uint8_t spi_get_module(spi_t *obj);
+void spi_async_free_handle(spi_async_handle_t* handle);
 
-/**@}*/
-
-#if DEVICE_SPI_ASYNCH
 /**
- * \defgroup AsynchSPI Asynchronous SPI Hardware Abstraction Layer
- * @{
+ * Frees the SPI instance.
```

Comment:
Do we want to specify what does it mean? eg should it be clocked out?

Reply:
I am not sure how clocking should be defined as if you only use masters then you could even clock the peripheral out between transaction where you might want to keep it enabled in slave mode.
Unless slave selection is handled externally to this API and then we can rely 100% on the spi_transfer to manage clocking of the peripheral.